# Test Wiki backlinks coverage
We've already collected the backlink counts [here](get_geonames_wiki_backlink_count.py) (stored [here](/hg190/corpora/GeoNames/wiki_backlink_counts.tsv)), so let's see how much of GeoNames is covered by this data.

This will help us determine if we need to extend the coverage to smaller toponyms (e.g. streets) with a back-off strategy: e.g. get the backlinks for a toponym's containing city, state, country, etc.

In [2]:
import pandas as pd
from data_helpers import load_simplified_geonames_data, load_geocorpora_data

## Load data

In [74]:
backlinks_data = pd.read_csv('/hg190/corpora/GeoNames/wiki_backlink_counts.tsv', sep='\t', index_col=False, encoding='utf-8')
geonames_data = load_simplified_geonames_data()

In [75]:
geonames_data.head()

geonames_ID                     name  \
0      2986043       Pic de Font Blanca   
1      2994701                 Roc Mélé   
2      3007683     Pic des Langounelles   
3      3017832     Pic de les Abelletes   
4      3017833  Estany de les Abelletes   

                                     alternate_names  latitude  longitude  \
0                     Pic de Font Blanca,Pic du Port  42.64991    1.53335   
1                        Roc Mele,Roc Meler,Roc Mélé  42.58765    1.74028   
2                               Pic des Langounelles  42.61203    1.47364   
3  Pic de la Font-Negre,Pic de la Font-Nègre,Pic ...  42.52535    1.73343   
4  Estany de les Abelletes,Etang de Font-Negre,Ét...  42.52915    1.73362   

  feature_class feature_code  population  alternate_name_count  
0             T           PK           0                     2  
1             T           MT           0                     3  
2             T           PK           0                     1  
3             T           PK           0                     3  
4             H           LK           0                     3

In [76]:
# merge with geonames data
geonames_data = pd.merge(geonames_data, backlinks_data.loc[:, ['backlink_count', 'geoname_ID']], 
                         left_on='geonames_ID', right_on='geoname_ID', how='outer')
geonames_data.drop('geoname_ID', axis=1, inplace=True)
geonames_data.head()

geonames_ID                     name  \
0    2986043.0       Pic de Font Blanca   
1    2994701.0                 Roc Mélé   
2    3007683.0     Pic des Langounelles   
3    3017832.0     Pic de les Abelletes   
4    3017833.0  Estany de les Abelletes   

                                     alternate_names  latitude  longitude  \
0                     Pic de Font Blanca,Pic du Port  42.64991    1.53335   
1                        Roc Mele,Roc Meler,Roc Mélé  42.58765    1.74028   
2                               Pic des Langounelles  42.61203    1.47364   
3  Pic de la Font-Negre,Pic de la Font-Nègre,Pic ...  42.52535    1.73343   
4  Estany de les Abelletes,Etang de Font-Negre,Ét...  42.52915    1.73362   

  feature_class feature_code  population  alternate_name_count  backlink_count  
0             T           PK         0.0                   2.0             NaN  
1             T           MT         0.0                   3.0             NaN  
2             T           PK         0.0                   1.0             NaN  
3             T           PK         0.0                   3.0             NaN  
4             H           LK         0.0                   3.0             NaN

In [77]:
geonames_data.loc[:, 'geonames_ID'] = geonames_data.loc[:, 'geonames_ID'].fillna(0, inplace=False).astype(int)

In [78]:
# replace nans
geonames_data.loc[:, 'backlink_count'] = geonames_data.loc[:, 'backlink_count'].fillna(-1, inplace=False)

In [79]:
print('%d/%d GeoNames entries have Wiki data'%
      (geonames_data[geonames_data.loc[:, 'backlink_count'] >= 0].shape[0], geonames_data.shape[0]))

525329/11686810 GeoNames entries have Wiki data


So it looks like 5% of all the Geonames entries have Wiki data, which seems extremely sparse.

We will need to extend the counts.

What kinds of entries do not have Wiki data?

In [80]:
print('blah')

blah


In [81]:
geonames_data.shape

(11686810, 10)

In [83]:
pd.np.random.seed(123)
non_wiki_data = geonames_data[geonames_data.loc[:, 'backlink_count'] == -1]
# shuffle data
non_wiki_data = non_wiki_data.loc[pd.np.random.choice(non_wiki_data.index, size=non_wiki_data.shape[0], replace=False), :]
non_wiki_data.head(100)

geonames_ID                                               name  \
1459317       1555591                                           Tianqiao   
7495944       1334926                                         Mirza Hill   
4071142      10603180                                            Bargāon   
7639191      11095983                                    Shambaliak Jhal   
10980560      7154846            Dillon County Fire Department Station 1   
8188035       8059397                                            Maloyuz   
11113686      7644832  Holiday Inn Express Ft. Lauderdale Airport / C...   
11244183      9847280                          Best Western Bayfront Inn   
1261732        235820                                              Zigbo   
566461        3294256                                           Bugojuše   
5283677       1237511                                          Mabulgoda   
1851503       9901038                                           Hengtang   
1903350       9999567                                       Zhuanzhilian   
395571        8193768                                       Waterloo Bay   
5179622       1483717                                      Gora Beskesik   
2383871       2823194                                               Thal   
7205053       6230261                                          Teheoriri   
1388039       3871072                                       Sección Lazo   
11297872     11778219                                  South Beacon Hill   
1148101       6148851                                       Skiakl Point   
9474739       4456241                                    Blackwoods Lake   
11326844      3627788                                     Hato San Pedro   
2872127        661997                                               Alho   
11085435      7264976                                    Interchange 343   
9566132       4548021                                         Pine Creek   
7612988       7505143                                Goth Chaudhry Yūsuf   
3388513       1626239                                     Pematang Sulah   
431011        8793227                                         Cooeeville   
4218245      10786069                                       Kotta Kunkām   
102939       11110538                                                Hin   
...               ...                                                ...   
8078212       2116020                                  Klyuch Kultuchnyy   
3278480      11486787                                    Panaxit Segundo   
319899        2142876                                       Wowingragong   
5717431        934681                                       River Boucan   
3669860       7610610                                         Malapedo A   
160214        7858819                                             Mbuale   
7514430       1357503                                               Lang   
1456940       1553212                                        Qinggangcun   
10084511      5077321                        Saint Pauls Lutheran Church   
1264425        238527                                               Kada   
2783174       6357101                               Alcolea de Calatrava   
5893317       8784459                                    Arroyo el Junco   
1759946       8517101                                            Luozhen   
11551423       963560                                       Quarie Kloof   
7923958        523965                               Stantsiya Myansel’ga   
7402529       1698272                                             Nangka   
11530978       943103                                 Vogelstruisfontein   
5094705       6684559                                        Dongmakdong   
4409640      11660784                                      Vellavāypudūr   
10611437      5606136                                  Sage Creek School   
10781853      5777314                         Lincoln Juni

Lots of small toponyms!

- `Maloyuz` => stream
- `Best Western Bayfront Inn` => hotel
- `Interchange 343` => highway exit
- `Qinggangcun` => village?
- `Gora Beskesik` => hill

## Extend backlink counts
The original GeoNames data has administrative and country containment, which we can use to extend the backlink counts to smaller entities.

TODO: doing this naive extension still won't help all comparison cases, such as cases where the levels don't match up: e.g., if TOPO_A has both admin4 and admin1 Wiki data but TOPO_B only has admin1...do we weight the levels differently?

In [29]:
col_names = ['geonames_ID', 'country_code', 'cc2', 'admin1', 'admin2', 'admin3', 'admin4']
use_cols = [0, 8, 9, 10, 11, 12, 13]
geonames_admin_data = pd.read_csv('/hg190/corpora/GeoNames/allCountries.zip', sep='\t',
                                  compression='zip', 
                                  header=None, usecols=use_cols, names = col_names, 
                                  index_col=False, encoding='utf-8')
geonames_admin_data.columns = col_names
geonames_admin_data.head()

/hg190/istewart6/crisis_language/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (9,10,11,12,13) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


geonames_ID country_code    cc2 admin1 admin2 admin3 admin4
0      2986043           AD    NaN     00    NaN    NaN    NaN
1      2994701           AD  AD,FR     00    NaN    NaN    NaN
2      3007683           AD  AD,FR     00    NaN    NaN    NaN
3      3017832           AD     FR     A9     66    663  66146
4      3017833           AD     FR     A9    NaN    NaN    NaN

In [49]:
# join with names so we have an idea of what we're doing
geonames_admin_data = pd.merge(geonames_admin_data, geonames_data.loc[:, ['geonames_ID', 'name']], 
                               on='geonames_ID', how='inner')
geonames_admin_data.head(50)

geonames_ID country_code    cc2 admin1 admin2 admin3 admin4  \
0      2986043           AD    NaN     00    NaN    NaN    NaN   
1      2994701           AD  AD,FR     00    NaN    NaN    NaN   
2      3007683           AD  AD,FR     00    NaN    NaN    NaN   
3      3017832           AD     FR     A9     66    663  66146   
4      3017833           AD     FR     A9    NaN    NaN    NaN   
5      3023203           AD    NaN     00    NaN    NaN    NaN   
6      3029315           AD  AD,FR     B3     09     91   9139   
7      3034945           AD    NaN     00    NaN    NaN    NaN   
8      3038814           AD    NaN     07    NaN    NaN    NaN   
9      3038815           AD    NaN     04    NaN    NaN    NaN   
10     3038816           AD    NaN     04    NaN    NaN    NaN   
11     3038818           AD    NaN     02    NaN    NaN    NaN   
12     3038819           AD    NaN     08    NaN    NaN    NaN   
13     3038820           AD    NaN     04    NaN    NaN    NaN   
14     3038821           AD    NaN     04    NaN    NaN    NaN   
15     3038822           AD    NaN     08    NaN    NaN    NaN   
16     3038823           AD    NaN     07    NaN    NaN    NaN   
17     3038824           AD    NaN     06    NaN    NaN    NaN   
18     3038825           AD    NaN     06    NaN    NaN    NaN   
19     3038826           AD  AD,ES     00    NaN    NaN    NaN   
20     3038827           AD    NaN     02    NaN    NaN    NaN   
21     3038828           AD    NaN     03    NaN    NaN    NaN   
22     3038829           AD    NaN     04    NaN    NaN    NaN   
23     3038830           AD    NaN     05    NaN    NaN    NaN   
24     3038831           AD    NaN     03    NaN    NaN    NaN   
25     3038832           AD    NaN     03    NaN    NaN    NaN   
26     3038833           AD    NaN     06    NaN    NaN    NaN   
27     3038834           AD    NaN     02    NaN    NaN    NaN   
28     3038835           AD    NaN     04    NaN    NaN    NaN   
29     3038836           AD    NaN     04    NaN    NaN    NaN   
30     3038837           AD    NaN     05    NaN    NaN    NaN   
31     3038838           AD    NaN     08    NaN    NaN    NaN   
32     3038839           AD    NaN     08    NaN    NaN    NaN   
33     3038840           AD    NaN     06    NaN    NaN    NaN   
34     3038841           AD    NaN     05    NaN    NaN    NaN   
35     3038842           AD    NaN     03    NaN    NaN    NaN   
36     3038843           AD    NaN     06    NaN    NaN    NaN   
37     3038844           AD    NaN     00    NaN    NaN    NaN   
38     3038845           AD     ES     00    NaN    NaN    NaN   
39     3038846           AD    NaN     04    NaN    NaN    NaN   
40     3038847           AD    NaN     04    NaN    NaN    NaN   
41     3038848           AD    NaN     04    NaN    NaN    NaN   
42     3038849           AD    NaN     04    NaN    NaN    NaN   
43     3038850           AD    NaN     06    NaN    NaN    NaN   
44     3038851           AD    NaN     04    NaN    NaN    NaN   
45     3038852           AD    NaN     05    NaN    NaN    NaN   
46     3038853           AD    NaN     05    NaN    NaN    NaN   
47     3038854           AD    NaN     04    NaN    NaN    NaN   
48     3038855           AD    NaN     05    NaN    NaN    NaN   
49     3038856           AD    NaN     00    NaN    NaN    NaN   

                            name  
0             Pic de Font Blanca  
1                       Roc Mélé  
2           Pic des Langounelles  
3           Pic de les Abelletes  
4        Estany de les Abelletes  
5   Port Vieux de la Coume d’Ose  
6           Port de la Cabanette  
7                      Port Dret  
8                Costa de Xurius  
9                Font de la Xona  
10                     Xixerella  
11                       Riu Xic  
12                   Pas del Xic  
13                  Roc del Xeig  
14              Collada del Xeig  
15                   Fonts Vives  
16                  Roc de Vista  
17    

In [52]:
geonames_admin_data.iloc[10000:10050, :]

geonames_ID country_code  cc2 admin1 admin2 admin3 admin4  \
10000     1121064           AF  NaN     26    NaN    NaN    NaN   
10001     1121065           AF  NaN     09    NaN    NaN    NaN   
10002     1121066           AF  NaN     09    NaN    NaN    NaN   
10003     1121067           AF  NaN     41    NaN    NaN    NaN   
10004     1121068           AF  NaN     39    NaN    NaN    NaN   
10005     1121069           AF  NaN     26    NaN    NaN    NaN   
10006     1121070           AF  NaN     10    NaN    NaN    NaN   
10007     1121071           AF  NaN     30    NaN    NaN    NaN   
10008     1121072           AF  NaN     11    NaN    NaN    NaN   
10009     1121073           AF  NaN     37    NaN    NaN    NaN   
10010     1121074           AF  NaN     06    NaN    NaN    NaN   
10011     1121075           AF  NaN     23    NaN    NaN    NaN   
10012     1121076           AF  NaN     10    NaN    NaN    NaN   
10013     1121077           AF  NaN     03    NaN    NaN    NaN   
10014     1121078           AF  NaN     17    NaN    NaN    NaN   
10015     1121079           AF  NaN     10    NaN    NaN    NaN   
10016     1121080           AF  NaN     11    NaN    NaN    NaN   
10017     1121081           AF  NaN     02    NaN    NaN    NaN   
10018     1121082           AF  NaN     06    NaN    NaN    NaN   
10019     1121083           AF  NaN     29    NaN    NaN    NaN   
10020     1121084           AF  NaN     02    NaN    NaN    NaN   
10021     1121085           AF  NaN     28    NaN    NaN    NaN   
10022     1121086           AF  NaN     06    NaN    NaN    NaN   
10023     1121087           AF  NaN     09    NaN    NaN    NaN   
10024     1121088           AF  NaN     29    NaN    NaN    NaN   
10025     1121089           AF  NaN     23    NaN    NaN    NaN   
10026     1121090           AF  NaN     40    NaN    NaN    NaN   
10027     1121091           AF  NaN     28    NaN    NaN    NaN   
10028     1121092           AF  NaN     11    NaN    NaN    NaN   
10029     1121093           AF  NaN     28    NaN    NaN    NaN   
10030     1121094           AF  NaN     08    NaN    NaN    NaN   
10031     1121095           AF  NaN     24    NaN    NaN    NaN   
10032     1121096           AF  NaN     24    NaN    NaN    NaN   
10033     1121097           AF  NaN     17    NaN    NaN    NaN   
10034     1121098           AF  NaN     18    NaN    NaN    NaN   
10035     1121099           AF  NaN     09    NaN    NaN    NaN   
10036     1121100           AF  NaN     09    NaN    NaN    NaN   
10037     1121101           AF  NaN     23    NaN    NaN    NaN   
10038     1121102           AF  NaN     27    NaN    NaN    NaN   
10039     1121103           AF  NaN     11    NaN    NaN    NaN   
10040     1121104           AF  NaN     37    NaN    NaN    NaN   
10041     1121105           AF  NaN     08    NaN    NaN    NaN   
10042     1121106           AF  NaN     02    NaN    NaN    NaN   
10043     1121107           AF  NaN     02    NaN    NaN    NaN   
10044     1121108           AF  NaN     31    NaN    NaN    NaN   
10045     1121109           AF  NaN     32    NaN    NaN    NaN   
10046     1121110           AF  NaN     06    NaN    NaN    NaN   
10047     1121111           AF  NaN     23    NaN    NaN    NaN   
10048     1121112           AF  NaN     28    NaN    NaN    NaN   
10049     1121113           AF  NaN     24    NaN    NaN    NaN   

                       name  
10000              Zamch Āb  
10001  Darah-ye Zambūr Khān  
10002           Zanbūr Khān  
10003             Zanbūrkhū  
10004              Zambūray  
10005              Zanbūrak  
10006               Zamboli  
10007              Zambukan  
10008             Zamīn Kaj  
10009                Zambar  
10010          Kōh-e Zamzam  
10011           Zamān Nāwêr  
10012          Zamān Māndêh  
10013            Zamān Khēl  
10014            Zamān Khēl  
10015      Zamān Khān Kêlay  
10016             Zamānābād  
10017                 Zamān  
10018             

Notes:

- `00` as the admin1 region means that it cannot be resolved to an administrative region.
- `NaN` as the cc2 code means that there is no dispute over country? Not sure.

First: collect the backlink count for all available admin regions.

In [58]:
# country
country_data_cols = ['ISO', 'Population', 'geonameid']
country_data = pd.read_csv('/hg190/corpora/GeoNames/countryInfo.txt', sep='\t', comment='#', usecols=country_data_cols)
# admin 1
admin_1_data_cols = [0, 3]
admin_1_data_col_names = ['country.admin1', 'geonames_ID']
admin_1_data = pd.read_csv('/hg190/corpora/GeoNames/admin1CodesASCII.txt', sep='\t', 
                          header=None, usecols=admin_1_data_cols, names=admin_1_data_col_names)
# admin 2
admin_2_data_cols = [0, 3]
admin_2_data_col_names = ['country.admin2.geonames_ID', 'geonames_ID']
admin_2_data = pd.read_csv('/hg190/corpora/GeoNames/admin2Codes.txt', sep='\t', 
                           header=None, usecols=admin_2_data_cols, names=admin_2_data_col_names)
# admin 3

# admin 4

In [61]:
admin_1_data.head(20)

country.admin1  geonames_ID
0           AD.06      3039162
1           AD.05      3039676
2           AD.04      3040131
3           AD.03      3040684
4           AD.02      3041203
5           AD.07      3041566
6           AD.08      3338529
7           AE.07       290595
8           AE.05       291075
9           AE.03       292224
10          AE.06       292673
11          AE.04       292879
12          AE.02       292933
13          AE.01       292969
14          AF.28      1121143
15          AF.27      1121863
16          AF.26      1123230
17          AF.33      1127106
18          AF.32      1127766
19          AF.40      1131054

In [63]:
test_country_admin1 = 'AF.01'
print(admin_1_data[admin_1_data.loc[:, 'country.admin1'] == test_country_admin1])

   country.admin1  geonames_ID
43          AF.01      1147745


In [65]:
admin_2_data.head(1000).tail(100)

country.admin2.geonames_ID  geonames_ID
900                AR.05.14147      3836128
901                AR.02.10091      3836129
902                AR.18.70098      3836188
903                AR.10.38077      3836444
904                AR.10.38070      3836473
905                AR.23.94007      3836551
906                AR.13.50105      3836668
907                AR.10.38063      3836752
908                AR.01.06763      3836848
909                AR.13.50098      3836982
910                AR.19.74063      3836983
911                AR.21.82126      3836984
912                AR.18.70091      3836985
913                AR.22.86175      3836986
914                AR.03.22147      3837081
915                AR.21.82119      3837090
916                AR.05.14140      3837121
917                AR.21.82112      3837122
918                AR.21.82105      3837460
919                AR.05.14133      3837475
920                AR.21.82091      3837821
921                AR.13.50091      3837885
922                AR.17.66154      3837886
923                AR.12.46119      3837918
924                AR.16.62077      3838028
925                AR.10.38056      3838029
926                AR.05.14126      3838154
927                AR.12.46126      3838186
928                AR.01.06714      3838222
929                AR.01.06721      3838240
..                         ...          ...
970                AR.15.58098      3841298
971                AR.15.58105      3841303
972                AR.16.62063      3841356
973                AR.01.06623      3841488
974                AR.01.06616      3841610
975                AR.22.86133      3841612
976                AR.15.58091      3841673
977                AR.01.06609      3841678
978                AR.01.06602      3841799
979                AR.04.26070      3841892
980                AR.02.10077      3842418
981                AR.17.66126      3842606
982                AR.01.06595      3842669
983                AR.22.86126      3842719
984                AR.03.22112      3842751
985                AR.01.06588      3842871
986                AR.03.22105      3842872
987                AR.16.62049      3842875
988                AR.18.70063      3842876
989                AR.21.82077      3842877
990                AR.16.62056      3842988
991                AR.15.58084      3842991
992                AR.22.86119      3843684
993                AR.24.90070      3843802
994                AR.17.66119      3843991
995                AR.22.86112      3844110
996                AR.15.58077      3844208
997                AR.05.14070      3844209
998                AR.17.66112      3844297
999                AR.03.22098      3844638

[100 rows x 2 columns]

In [ ]:
# ordered admin columns from smallest => largest
ORDERED_ADMINS = ['admin4', 'admin3', 'admin2', 'admin1']
def expand_data(geo_id, admin_data):
    """
    For a given ID without data,
    we assign it data by starting at the
    lowest-possible administration region and backing
    off until we get a hit.
    """
    g_data = admin_data[admin_data.loc[:, 'geonames_ID'] == geo_id].iloc[0, :]
    